src/main/java/io/quarkus/gizmo/ForEachLoopImpl.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2022 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.gizmo;
+
+class ForEachLoopImpl extends LoopImpl implements ForEachLoop {
+
+    private final ResultHandle element;
+    private final BranchResult result;
+
+    ForEachLoopImpl(BytecodeCreatorImpl enclosing, ResultHandle iterable) {
+        super(enclosing);
+        ResultHandle iterator = Gizmo.invokeIterable(enclosing).instance(iterable).iterator();
+        this.result = ifTrue(Gizmo.invokeIterator(this).instance(iterator).hasNext());
+        this.element = Gizmo.invokeIterator(result.trueBranch()).instance(iterator).next();
+    }
+
+    @Override
+    public ResultHandle element() {
```

Comment:
> which means users calling forLoop.element() more than once will get two iterations

I don't think so. This `element` is just the result of `it.next()` that can be used in the `block()`. You can call `element()` as many times as you want but you always get the same reference. Note that you're only generating the code that does iteration. But you don't iterate...

Reply:
I really don't think it works like that. Can you check the resulting bytecode to be sure? Seems to me that `element` is literally a call to `foo.next()` so if you call it more than once per iteration you will get that called more than once, which is dangerous.

Reply:
Nope, it's not. If you duplicate [this line](https://github.com/quarkusio/gizmo/pull/115/files#diff-cc1aaa93f196abeaba7c90a119a54cd226d8972aa58083ed8c38a181feab4f48R27) you'll get the result `foofoo-barbar-` because the generated (decompiled) bytecode is like:
```java
public class MyTest implements Supplier {
	@Override
	public Object get() {
		final StringBuilder sb = new StringBuilder();
		for (final Object next : List.of("foo", "bar")) {
			sb.append(next.toString());
			sb.append(next.toString());
			sb.append("-");
		}
		return sb.toString();
	}
}
```

Note that `element` is final and initialized in the constructor of the `ForEachLoopImpl`.

Reply:
I know it's final and initialised in the constructor, but it's not like it's declaring a variable. I don't understand why this works. IMO the `element` only contains instructions to call `iter.next()` so inserting it twice in the tree should call it twice. When you want a local variable, gizmo has a different API. This really is the decompiled bytecode you tested?

Reply:
But, what's the point of `AssignableResultHandle` if we can reference the result multiple times, then?

Reply:
This is correct, ResultHandle is generally conceptually a variable. What complicates it here is that it is actually one variable per iteration. Basically the code that is written in the constructor is executed at the start of every iteration.

Reply:
> But, what's the point of `AssignableResultHandle` if we can reference the result multiple times, then?

`AssignableResultHandle` can be reassigned whereas `ResultHandle` is more like a `final` local variable...?